cpp/src/binaryop/compiled/equality_ops.cu

```diff
@@ -14,44 +14,75 @@
+void dispatch_equality_op(mutable_column_view& out,
+                          column_view const& lhs,
+                          column_view const& rhs,
                           bool is_lhs_scalar,
                           bool is_rhs_scalar,
                           binary_operator op,
                           rmm::cuda_stream_view stream)
 {
-  auto common_dtype = get_common_type(outd.type(), lhsd.type(), rhsd.type());
-
-  // Execute it on every element
+  if (is_struct(lhs.type()) && is_struct(rhs.type())) {
+    auto const nullability =
+      structs::detail::contains_null_structs(lhs) || structs::detail::contains_null_structs(rhs)
+        ? structs::detail::column_nullability::FORCE
+        : structs::detail::column_nullability::MATCH_INCOMING;
+    auto const lhs_flattened =
+      structs::detail::flatten_nested_columns(table_view{{lhs}}, {}, {}, nullability);
+    auto const rhs_flattened =
+      structs::detail::flatten_nested_columns(table_view{{rhs}}, {}, {}, nullability);
+    auto d_lhs = table_device_view::create(lhs_flattened);
+    auto d_rhs = table_device_view::create(rhs_flattened);
 
-  if (common_dtype) {
-    if (op == binary_operator::EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_device_dispatcher<ops::Equal>{
-                 *common_dtype, outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
-    } else if (op == binary_operator::NOT_EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_device_dispatcher<ops::NotEqual>{
-                 *common_dtype, outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
-    }
+    if (op == binary_operator::EQUAL || op == binary_operator::NOT_EQUAL)
+      detail::struct_compare(
+        out,
+        row_equality_comparator{
+          nullate::DYNAMIC{has_nested_nulls(lhs_flattened) || has_nested_nulls(rhs_flattened)},
+          *d_lhs,
+          *d_rhs},
+        is_lhs_scalar,
+        is_rhs_scalar,
+        op == binary_operator::NOT_EQUAL,
+        stream);
+    else
+      CUDF_FAIL("Unsupported operator for these types");
   } else {
-    if (op == binary_operator::EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_double_device_dispatcher<ops::Equal>{
-                 outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
-    } else if (op == binary_operator::NOT_EQUAL) {
-      for_each(stream,
-               outd.size(),
-               binary_op_double_device_dispatcher<ops::NotEqual>{
-                 outd, lhsd, rhsd, is_lhs_scalar, is_rhs_scalar});
+    auto common_dtype = get_common_type(out.type(), lhs.type(), rhs.type());
+    auto outd         = mutable_column_device_view::create(out, stream);
+    auto lhsd         = column_device_view::create(lhs, stream);
+    auto rhsd         = column_device_view::create(rhs, stream);
+    if (common_dtype) {
+      if (op == binary_operator::EQUAL) {
```

Comment:
I wish this to be simpler.
We should split this into 2 files in another PR. (might  increase in compile time (+20s), but it will become easy to maintain)